spark-cassandra-connector/scala-2.11/src/main/scala/com/datastax/spark/connector/rdd/reader/AnyObjectFactory.scala

```diff
@@ -0,0 +1,194 @@
+package com.datastax.spark.connector.rdd.reader
+
+import java.lang.reflect.Constructor
+
+import com.datastax.spark.connector.util.Logging
+import com.google.common.primitives.Primitives
+import com.thoughtworks.paranamer.AdaptiveParanamer
+import org.apache.commons.lang3.reflect.ConstructorUtils
+
+import scala.reflect.runtime.universe._
+import scala.util.{Failure, Success, Try}
+
+/** Factory for creating objects of any type by invoking their primary constructor.
+  * Unlike Java reflection Methods or Scala reflection Mirrors, this factory is serializable
+  * and can be safely passed along with Spark tasks. */
+class AnyObjectFactory[T: TypeTag] extends Logging with Serializable {
+
+  import AnyObjectFactory._
+
+  @transient
+  private val tpe = implicitly[TypeTag[T]].tpe
+
+  @transient
+  lazy val rm: RuntimeMirror =
+    runtimeMirror(Thread.currentThread().getContextClassLoader)
+
+  // This must be serialized:
+  val javaClass: Class[T] =
+    rm.runtimeClass(tpe).asInstanceOf[Class[T]]
+
+  // This must be serialized:
+  val constructorParamTypeNames: IndexedSeq[ParamType] =
+    toParamTypeNames(resolveConstructor(javaClass))
+
+  @transient
+  private lazy val javaConstructor: Constructor[T] =
+    resolveConstructorFromParamTypeNames(javaClass, constructorParamTypeNames)
+
+  // It is quite important to invoke constructorParamTypes here because it invokes client side validation whether
+  // the right constructor exists or not, before the job is started
+  val argCount: Int = constructorParamTypes.length
+
+  val argOffset: Int = oneIfMemberClass(javaClass)
+
+  @transient
+  lazy val constructorParamTypes: Array[Type] = {
+    val requiredParamClasses = javaConstructor.getParameterTypes
+      .drop(AnyObjectFactory.oneIfMemberClass(javaClass))
+
+    tpe.decl(termNames.CONSTRUCTOR).asTerm.alternatives.map { term =>
```

Comment:
Yes, please extract it to ReflectionUtil. 
The same applies to ClassBasedRowReader. There is no point in forking the whole class if the only thing that requires special treating is a well defined operation for getting the constructor definition.